README.md
# vpn cookbook
Install and configure [OpenVPN](https://openvpn.net/).

## Usage

### OpenVPN server

Install OpenVPN server packages, configure a Certification Authority and launch the service.

```ruby
vpn_server 'lan' do
  fqdn 'vpn.acme.corp'
  user 'vagrant'
  group 'vagrant'
  certificate(
    country: 'RU',
    province: 'Samara',
    city: 'Samara',
    org: 'ACME Corp.',
    email: 'admin@acme.corp',
    ou: 'IT'
  )
  port 1194
  network '10.1.0.0/24'
  openvpn(
    proto: 'tcp',
    cipher: 'AES-128-CBC'
  )
action :setup
end
```

In this particular case, the service name is `openvpn@lan`.

### OpenVPN client

Configure an OpenVPN client with a static IPv4 address.

```ruby
vpn_client 'Alice VPN connection' do
  name 'alice'
  user 'vagrant'
  group 'vagrant'
  server 'lan'
  ipv4_address '10.1.0.3'
  action :create
end
```

In this particular case `*.ovpn` configuration files are stored in `/etc/chef-vpn/client-config/lan/files` directory.

## License
MIT © [Alexander Pyatkin](https://github.com/aspyatkin)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2020-07-04
### Added
- Support Ubuntu 20.04.

## [0.3.0] - 2020-01-02
### Added
- Add `redirect_gateway`, `bypass_dhcp` and `bypass_dns` options.

## [0.2.3] - 2019-04-19

### Changed
- Update [line](https://supermarket.chef.io/cookbooks/line) cookbook dependency.

## [0.2.2] - 2019-04-13

### Fixed
- Fix a bug during subsequent Chef client runs.

## [0.2.1] - 2019-04-12

### Added
- Upload the first working version of the cookbook to [Chef Supermarket](https://supermarket.chef.io/cookbooks/vpn).

## [0.0.1] - 2018-11-26

Create the cookbook.


resources/client.rb
resource_name :vpn_client
property :name, String, name_property: true

property :basedir, String, default: '/etc/chef-vpn'
property :user, String, required: true
property :group, String, required: true
property :server, String, required: true
property :ipv4_address, String, required: true

default_action :create

action :create do
  ca_dir_path = ::File.join(new_resource.basedir, 'server', new_resource.server)
  key_dir_name = 'keys'

  client_crt_file_name = "#{new_resource.name}.crt"
  client_key_file_name = "#{new_resource.name}.key"

  key_dir_path = ::File.join(ca_dir_path, key_dir_name)

  if node['platform'] == 'ubuntu' && node['platform_version'].to_i == 20
    client_crt_file_path = ::File.join(key_dir_path, 'issued', client_crt_file_name)
    client_key_file_path = ::File.join(key_dir_path, 'private', client_key_file_name)

    bash "generate client #{new_resource.name} key at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        ./easyrsa --batch build-client-full #{new_resource.name} nopass
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(client_crt_file_path) && ::File.exist?(client_key_file_path) }
    end
  else
    client_crt_file_path = ::File.join(key_dir_path, client_crt_file_name)
    client_key_file_path = ::File.join(key_dir_path, client_key_file_name)

    bash "generate client #{new_resource.name} key at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        source ./vars
        ./build-key --batch #{new_resource.name}
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(client_crt_file_path) && ::File.exist?(client_key_file_path) }
    end
  end

  client_config_dir_path = ::File.join(new_resource.basedir, 'client-config', new_resource.server)
  client_config_file_dir_path = ::File.join(client_config_dir_path, 'files')

  client_ovpn_file_name = "#{new_resource.name}.ovpn"
  client_ovpn_file_path = ::File.join(client_config_file_dir_path, client_ovpn_file_name)

  bash "generate client #{new_resource.name} config file at #{client_config_dir_path}" do
    cwd client_config_dir_path
    code <<-EOH
      ./make-config #{new_resource.name}
    EOH
    user new_resource.user
    group new_resource.group
    action :run
    not_if { ::File.exist?(client_ovpn_file_path) }
  end

  client_opts_file = ::File.join('/etc', 'openvpn', new_resource.server, 'clients', new_resource.name)

  require 'ip'

  template client_opts_file do
    cookbook 'vpn'
    source 'client_opts.erb'
    owner 'root'
    group node['root_group']
    variables(
      ipv4_address: new_resource.ipv4_address,
      network: ::IP.new(resources("vpn_server[#{new_resource.server}]").network)
    )
    mode 0o644
    action :create
  end
end


resources/server.rb
resource_name :vpn_server
property :name, String, name_property: true

property :basedir, String, default: '/etc/chef-vpn'
property :user, String, required: true
property :group, String, required: true

property :certificate, Hash, default: {}

property :key_size, Integer, default: 2048

property :fqdn, String, required: true
property :port, Integer, default: 1194
property :network, String, required: true
property :openvpn, Hash, default: {}

property :redirect_gateway, [TrueClass, FalseClass], default: false
property :bypass_dhcp, [TrueClass, FalseClass], default: true
property :bypass_dns, [TrueClass, FalseClass], default: true
property :manage_firewall_rules, [TrueClass, FalseClass], default: false
property :firewall_filter_rule_position, Integer, default: 50
property :firewall_nat_rule_position, Integer, default: 150

default_action :setup

action :setup do
  %w[
    openvpn
    easy-rsa
  ].each do |pkg_name|
    package pkg_name do
      action :install
    end
  end

  directory new_resource.basedir do
    owner new_resource.user
    group new_resource.group
    mode 0o755
    recursive true
    action :create
  end

  server_basedir = ::File.join(new_resource.basedir, 'server')

  directory server_basedir do
    owner new_resource.user
    group new_resource.group
    mode 0o755
    action :create
  end

  ca_dir_path = ::File.join(server_basedir, new_resource.name)

  execute "create CA directory at #{ca_dir_path}" do
    command "make-cadir #{ca_dir_path}"
    user new_resource.user
    group new_resource.group
    action :run
    not_if { ::File.directory?(ca_dir_path) }
  end

  key_dir_name = 'keys'
  key_name = 'server'
  crt_details = new_resource.certificate

  ca_crt_file_name = 'ca.crt'
  ca_key_file_name = 'ca.key'

  server_crt_file_name = "#{key_name}.crt"
  server_key_file_name = "#{key_name}.key"

  key_dir_path = ::File.join(ca_dir_path, key_dir_name)
  vars_file_path = ::File.join(ca_dir_path, 'vars')

  if node['platform'] == 'ubuntu' && node['platform_version'].to_i == 20
    defaults = {
      'EASYRSA_REQ_COUNTRY' => crt_details.fetch(:country, 'US'),
      'EASYRSA_REQ_PROVINCE' => crt_details.fetch(:province, 'California'),
      'EASYRSA_REQ_CITY' => crt_details.fetch(:city, 'San Francisco'),
      'EASYRSA_REQ_ORG' => crt_details.fetch(:org, 'Acme Inc.'),
      'EASYRSA_REQ_EMAIL' => crt_details.fetch(:email, ''),
      'EASYRSA_REQ_OU' => crt_details.fetch(:ou, ''),
      'EASYRSA_REQ_CN' => crt_details.fetch(:ca, "#{key_name} CA"),
      'EASYRSA_PKI' => "$PWD/#{key_dir_name}",
      'EASYRSA_KEY_SIZE' => new_resource.key_size
    }

    defaults.each do |key, val|
      replace_or_add "adjust #{key} in #{vars_file_path}" do
        path vars_file_path
        pattern ::Regexp.new("^#set_var #{key}.*$")
        line "set_var #{key} #{val.is_a?(::Numeric) ? val.to_s : val.dump}"
        replace_only true
        ignore_missing false
        action :edit
      end
    end

    bash "init PKI directory at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        ./easyrsa init-pki
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if do
        ::Dir.exist?(key_dir_path) && (
          ::Gem::Version.new(RUBY_VERSION) >= ::Gem::Version.new('2.4.0') ?
            !::Dir.empty?(key_dir_path) : (::Dir.entries(key_dir_path).size != 2)
        )
      end
    end

    ca_crt_file_path = ::File.join(key_dir_path, ca_crt_file_name)
    ca_key_file_path = ::File.join(key_dir_path, 'private', ca_key_file_name)

    bash "build CA at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        ./easyrsa --batch build-ca nopass
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(ca_crt_file_path) && ::File.exist?(ca_key_file_path) }
    end

    server_crt_file_path = ::File.join(key_dir_path, 'issued', server_crt_file_name)
    server_key_file_path = ::File.join(key_dir_path, 'private', server_key_file_name)

    bash "build key server at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        ./easyrsa --batch build-server-full #{key_name} nopass
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(server_crt_file_path) && ::File.exist?(server_key_file_path) }
    end

    dh_key_file_name = 'dh.pem'
    dh_key_file_path = ::File.join(key_dir_path, dh_key_file_name)

    bash "build dh key at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        ./easyrsa gen-dh
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(dh_key_file_path) }
    end
  else
    defaults = {
      'KEY_COUNTRY' => crt_details.fetch(:country, 'US'),
      'KEY_PROVINCE' => crt_details.fetch(:province, 'California'),
      'KEY_CITY' => crt_details.fetch(:city, 'San Francisco'),
      'KEY_ORG' => crt_details.fetch(:org, 'Acme Inc.'),
      'KEY_EMAIL' => crt_details.fetch(:email, ''),
      'KEY_OU' => crt_details.fetch(:ou, ''),
      'KEY_NAME' => key_name,
      'KEY_DIR' => "$EASY_RSA/#{key_dir_name}",
      'KEY_SIZE' => new_resource.key_size
    }

    defaults.each do |key, val|
      replace_or_add "adjust #{key} in #{vars_file_path}" do
        path vars_file_path
        pattern ::Regexp.new("^export #{key}=.*$")
        line "export #{key}=#{val.is_a?(::Numeric) ? val.to_s : val.dump}"
        replace_only true
        ignore_missing false
        action :edit
      end
    end

    if node['platform'] == 'ubuntu' && node['platform_version'].to_f >= 18.04
      execute "Fix openssl.cnf at #{ca_dir_path}" do
        cwd ca_dir_path
        command 'cp ./openssl-1.0.0.cnf ./openssl.cnf'
        user new_resource.user
        group new_resource.group
        action :run
        not_if { ::File.exist?(::File.join(ca_dir_path, 'openssl.cnf')) }
      end
    end

    bash "clean CA key directory at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        source ./vars
        ./clean-all
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if do
        ::Dir.exist?(key_dir_path) && (
          ::Gem::Version.new(RUBY_VERSION) >= ::Gem::Version.new('2.4.0') ?
            !::Dir.empty?(key_dir_path) : (::Dir.entries(key_dir_path).size != 2)
        )
      end
    end

    ca_crt_file_path = ::File.join(key_dir_path, ca_crt_file_name)
    ca_key_file_path = ::File.join(key_dir_path, ca_key_file_name)

    bash "initialize CA at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        source ./vars
        ./build-ca --batch
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(ca_crt_file_path) && ::File.exist?(ca_key_file_path) }
    end

    server_crt_file_path = ::File.join(key_dir_path, server_crt_file_name)
    server_key_file_path = ::File.join(key_dir_path, server_key_file_name)

    bash "build key server at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        source ./vars
        ./build-key-server --batch #{key_name}
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(server_crt_file_path) && ::File.exist?(server_key_file_path) }
    end

    dh_key_file_name = "dh#{new_resource.key_size}.pem"
    dh_key_file_path = ::File.join(key_dir_path, dh_key_file_name)

    bash "build dh key at #{ca_dir_path}" do
      cwd ca_dir_path
      code <<-EOH
        source ./vars
        ./build-dh
      EOH
      user new_resource.user
      group new_resource.group
      action :run
      not_if { ::File.exist?(dh_key_file_path) }
    end
  end

  ta_key_file_name = 'ta.key'
  ta_key_file_path = ::File.join(key_dir_path, ta_key_file_name)

  bash "build ta key at #{ca_dir_path}" do
    cwd ca_dir_path
    code <<-EOH
      openvpn --genkey --secret #{key_dir_name}/#{ta_key_file_name}
    EOH
    user new_resource.user
    group new_resource.group
    action :run
    not_if { ::File.exist?(ta_key_file_path) }
  end

  openvpn_basedir = ::File.join('/etc', 'openvpn')

  openvpn_service_basedir = ::File.join(openvpn_basedir, new_resource.name)

  directory openvpn_service_basedir do
    owner 'root'
    group node['root_group']
    mode 0o755
    action :create
  end

  client_dir = ::File.join(openvpn_service_basedir, 'clients')

  directory client_dir do
    owner 'root'
    group node['root_group']
    mode 0o755
    action :create
  end

  [
    ca_crt_file_path,
    ca_key_file_path,
    server_crt_file_path,
    server_key_file_path,
    dh_key_file_path,
    ta_key_file_path
  ].each do |file_path|
    file ::File.join(openvpn_service_basedir, ::File.basename(file_path)) do
      content(lazy { ::IO.read(file_path) })
      owner 'root'
      group node['root_group']
      mode 0o600
      sensitive true
      action :create
    end
  end

  service_conf_file_path = ::File.join(openvpn_basedir,
                                       "#{new_resource.name}.conf")

  require 'ip'

  openvpn_conf = new_resource.openvpn
  openvpn_proto = openvpn_conf.fetch(:proto, 'udp')
  openvpn_cipher = openvpn_conf.fetch(:cipher, 'AES-256-CBC')
  openvpn_auth = openvpn_conf.fetch(:auth, 'SHA256')

  log_dir = ::File.join('/var', 'log', 'openvpn')

  directory log_dir do
    owner 'root'
    group node['root_group']
    mode 0o755
    action :create
  end

  template service_conf_file_path do
    cookbook 'vpn'
    source 'server.conf.erb'
    owner 'root'
    group node['root_group']
    variables(
      port: new_resource.port,
      proto: openvpn_proto,
      ta_key: ::File.join(new_resource.name, ta_key_file_name),
      cipher: openvpn_cipher,
      auth: openvpn_auth,
      ca: ::File.join(new_resource.name, ca_crt_file_name),
      cert: ::File.join(new_resource.name, server_crt_file_name),
      key: ::File.join(new_resource.name, server_key_file_name),
      dh: ::File.join(new_resource.name, dh_key_file_name),
      network: ::IP.new(new_resource.network),
      ipp_file: ::File.join(new_resource.name, 'ipp.txt'),
      client_dir: ::File.join(new_resource.name, 'clients'),
      status_file: ::File.join(log_dir, "#{new_resource.name}-status.log"),
      redirect_gateway: new_resource.redirect_gateway,
      bypass_dhcp: new_resource.bypass_dhcp,
      bypass_dns: new_resource.bypass_dns
    )
    mode 0o644
    action :create
  end

  service "openvpn@#{new_resource.name}" do
    action %i[enable start]
  end

  if new_resource.manage_firewall_rules
    with_run_context :root do
      firewall_rule "openvpn-#{new_resource.name}" do
        port new_resource.port
        source '0.0.0.0/0'
        protocol openvpn_proto.to_sym
        position new_resource.firewall_filter_rule_position
        command :allow
      end

      if new_resource.redirect_gateway
        firewall_rule "openvpn-#{new_resource.name}-postrouting" do
          source new_resource.network  # this line is needed to bypass setting raw IPv6 rule
          raw lazy { "-A POSTROUTING -s #{new_resource.network} -o #{::ChefCookbook::VPN.public_interface} -j MASQUERADE" }
          position new_resource.firewall_nat_rule_position
          command :allow
        end
      end
    end
  end

  client_config_basedir = ::File.join(new_resource.basedir, 'client-config')

  directory client_config_basedir do
    owner new_resource.user
    group new_resource.group
    mode 0o755
    action :create
  end

  client_config_dir_path = ::File.join(client_config_basedir, new_resource.name)

  directory client_config_dir_path do
    owner new_resource.user
    group new_resource.group
    mode 0o755
    action :create
  end

  client_config_file_dir_path = ::File.join(client_config_dir_path, 'files')

  directory client_config_file_dir_path do
    owner new_resource.user
    group new_resource.group
    mode 0o700
    action :create
  end

  base_client_conf_file_name = 'base.conf'
  base_client_conf_file_path = ::File.join(client_config_dir_path,
                                           base_client_conf_file_name)

  template base_client_conf_file_path do
    cookbook 'vpn'
    source 'client.conf.erb'
    owner new_resource.user
    group new_resource.group
    variables(
      fqdn: new_resource.fqdn,
      port: new_resource.port,
      proto: openvpn_proto,
      cipher: openvpn_cipher,
      auth: openvpn_auth
    )
    mode 0o644
    action :create
  end

  make_config_script_path = ::File.join(client_config_dir_path, 'make-config')

  template make_config_script_path do
    cookbook 'vpn'
    source (node['platform'] == 'ubuntu' && node['platform_version'].to_i == 20) ? 'make-config-ubuntu-20.04.sh.erb' : 'make-config.sh.erb'
    owner new_resource.user
    group new_resource.group
    variables(
      key_dir: key_dir_path,
      output_dir: client_config_file_dir_path,
      base_config: base_client_conf_file_path
    )
    mode 0o700
    action :create
  end
end
